daml-lf/upgrades/src/main/scala/com/digitalasset/daml/lf/upgrades/Typecheck.scala

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package upgrades
+
+import com.daml.lf.data.Ref
+import com.daml.lf.language.Ast
+import scala.util.{Try, Success, Failure}
+import com.daml.lf.validation.AlphaEquiv
+
+case class Upgrading[A](past: A, present: A) {
+  def map[B](f: A => B): Upgrading[B] = Upgrading(f(past), f(present))
+  def fold[B](f: (A, A) => B): B = f(past, present)
+  def zip[B, C](that: Upgrading[B], f: (A, B) => C) =
+    Upgrading(f(this.past, that.past), f(this.present, that.present))
+}
+
+case class UpgradeError(message: String) extends Throwable(message)
```

Comment:
There is usually no good reason to why to extend `Throwable` directly instead of `RuntimeException`. Furthermore, since these errors merely wrap messages, I'd go with just returning an `Either[String, Unit]` instead of Exception-based results. Then the caller can decide how to wrap these error messages.

Reply:
Will swap out Try for Either.

These errors aren't normally strings, those are placeholders for the datatypes being fully expanded. Doing that now.

Moving this into validation, so I'll make these a validation error.

Reply:
fixed - I've stuck with try seeing as use of ValidationError in Typing uses thrown errors, so we have precedent for using stack traces.